src/ripple/app/tx/impl/SetTrust.cpp

```diff
@@ -317,9 +317,14 @@ SetTrust::doApply ()
         std::uint32_t const uFlagsIn (sleRippleState->getFieldU32 (sfFlags));
         std::uint32_t uFlagsOut (uFlagsIn);
 
-        if (bSetNoRipple && !bClearNoRipple && (bHigh ? saHighBalance : saLowBalance) >= zero)
+        if (bSetNoRipple && !bClearNoRipple)
         {
-            uFlagsOut |= (bHigh ? lsfHighNoRipple : lsfLowNoRipple);
+            if ((bHigh ? saHighBalance : saLowBalance) >= zero)
+                uFlagsOut |= (bHigh ? lsfHighNoRipple : lsfLowNoRipple);
+
+            else if (view().rules().enabled(fix1578))
```

Comment:
I'd prefer we structure this the other way:

```
if (view().rules().enabled(fix1578) && ((bHigh ? saHighBalance : saLowBalance) < zero))
    return tecNEGATIVE_BALANCE;
uFlagsOut |= (bHigh ? lsfHighNoRipple : lsfLowNoRipple);
```

As a sidenote, I really dislike this transactor. It needs to be improved.

Reply:
I agree that your restructuring is easier to think about, but I think it's transaction changing?  The interesting case is when the balance is negative and `fix1578` is not yet enabled.

- In the old code if the balance is negative then `uFlagsOut` is not touched.
- In the proposed restructure if `fix1578` is not enabled then the balance is not checked and `uFlagsOut` is modified.

I may be misreading.

And, yeah, I find this transactor really hard to think about.  I'm guessing it's designed to gripe as little as possible so when several changes are made if at least one of them is allowed then the changes that are allowed are made?  But that's just a guess.  If that's the model then my change violates it.

Reply:
Let's ignore the `bSetNoRipple && !bClearNoRipple` part. Let's focus only on the `(bHigh ? saHighBalance : saLowBalance) >= beast::zero` part; we'll call it X.

The old code does

```
if(X)
    something();
```

Your proposed change does:

```
if(X)
    something();
else if (fix1578)
    return tec;
```

Both cases will update `uFlagsOut` **iff** `X` is true; the new change, gated under `fix1578`, is to return an error **iff** `!X`. So far so good.

My code is:

```
if(fix1578 && !X)
    return tec;
something();
```

So you are right; my change is transaction-breaking outside the `fix1578` gate. :scream: 

I still believe that, as is, the code is more difficult to read than is necessary, but we can always fix that if this amendment is activated and we come back to remove the amendment gate. The code can then become:

```
if (!X)
    return tec;
something();
```